### ISSUE TITLE
GLTFExporter: Empty materials:[] array breaks in Oculus Home

### ISSUE THREAD

**Issue by marcocorriero** [mrdoob/three.js#15860]
##### Description of the problem

well, i tried to convert a animated model to add to my Oculus Home to a glb file, since it has something that blocks the home, it gives me back a corrupted copy/unanimated copy.

Fix please?

##### Three.js version

- [ ] Dev
- [ ] r101
- [x] ...

##### Browser

- [ ] All of them
- [x] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [ ] All of them
- [x] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by Mugen87** [comment]
Can you please describe in more detail how your workflow looks like? How exactly is your model animated? Via morph targets, skeletal animations or just with a bunch of keyframes? Is it possible for you to share the original model (before converting to `glTF`)?

**Comment by marcocorriero** [comment]
the model got extracted by windows (i love them , and some of them have multiple animations, but this one has just one animation.)

Here the original model before conversion (is a glb, but it needs to be re-converted again to a glb due to some stuff going in microsoft's models)

https://drive.google.com/open?id=1fQoZr9oeeE3mGuJMYH_DA5Xz-naEmBwx

**Comment by donmccurdy** [comment]
Cute model 😅

Could you share the code used to load and export the file? Note that animations are not "part of" the mesh itself, and you'll have to provide them to the exporter for them to be included. For example:

```js
exporter.parse( model, function (gltf) {
  // ...
}, { animations: animations } );
```

See: https://threejs.org/docs/#examples/exporters/GLTFExporter

The materials are missing because GLTFExporter does not yet support spec/gloss PBR materials (`KHR_materials_pbrSpecularGlossiness`).

**Comment by marcocorriero** [comment]
> Cute model 😅
> 
> Could you share the code used to load and export the file? Note that animations are not part of the mesh itself, and you'll have to provide them to the exporter for them to be included. For example:
> 
> ```js
> exporter.parse( model, function (gltf) {
>   // ...
> }, { animations: animations } );
> ```
> 
> See: https://threejs.org/docs/#examples/exporters/GLTFExporter
> 
> The materials are missing because GLTFExporter does not yet support spec/gloss PBR materials (`KHR_materials_pbrSpecularGlossiness`).

im not editing it, it just the https://threejs.org/editor/
even with static glb it just gives me back corrupted/not working GLB...

**Comment by donmccurdy** [comment]
Please be specific – "corrupted/not working" is pretty vague. I see two issues:

1. Model is not animated.
2. Materials are missing.

![screen shot 2019-03-03 at 6 16 44 pm](https://user-images.githubusercontent.com/1848368/53706741-9eefe000-3de0-11e9-808e-13b95f4b5f36.png)

The editor does not export animation, that is just not a feature that it has. A contributor has been working on support for playing animation in it recently, and perhaps export will be supported in the future.

The materials in this model are not exported because of https://github.com/mrdoob/three.js/issues/15351, which would be good to fix I think. 

If there are any other specific issues you are seeing, let us know thanks.

**Comment by marcocorriero** [comment]
> Please be specific – "corrupted/not working" is pretty vague. I see two issues:
> 
> 1. Model is not animated.
> 2. Materials are missing.
> 
> ![screen shot 2019-03-03 at 6 16 44 pm](https://user-images.githubusercontent.com/1848368/53706741-9eefe000-3de0-11e9-808e-13b95f4b5f36.png)
> 
> The editor does not export animation, that is just not a feature that it has. A contributor has been working on support for playing animation in it recently, and perhaps export will be supported in the future.
> 
> The materials in this model are not exported because of #15351, which would be good to fix I think.
> 
> If there are any other specific issues you are seeing, let us know thanks.

I’m trying to make it able to be imported on my oculus hone... is there a way to make it work with materials and animations?

**Comment by donmccurdy** [comment]
I cannot see what you are seeing in Oculus Home, I do not have a Rift. If there are problems other than the two I have mentioned, you will have to tell us what they are.

The features you're requesting are tracked in these issues:

- GLTFExporter: Support spec/gloss PBR materials (#15351)
- Would like animations on models imported in the editor included on scene export (#11642)

I suspect they'll both be fixed at some point, but I don't have an estimate. PRs would be welcome. In the meantime the Editor cannot support the features.

Closing as duplicate of the two issues above.

**Comment by marcocorriero** [comment]
> I cannot see what you are seeing in Oculus Home; I do not have a Rift. If there are problems other than the two I have mentioned, you will have to tell us what they are.
> 
> The features you're requesting are tracked in these issues:
> 
> * GLTFExporter: Support spec/gloss PBR materials (#15351)
> * Would like animations on models imported in the editor included on scene export (#11642)
> 
> I suspect they'll both be fixed at some point, but I don't know when. PRs would be welcome. In the meantime the Editor cannot support the features.
> 
> Closing as duplicate of the two issues above.

somehow it doesn't load the glb cuz is corrupted. No model , nothing , if someone makes a PR to fix three.js/editor would be really appreciated. Thanks!

**Comment by mrdoob** [comment]
I tried loading the .glb in the editor, then exporting it, then loading it again and I can see the model (minus the animations and material).

If the .glb doesn't load in Oculus home, you may want to report that to Oculus/Facebook instead.

**Comment by donmccurdy** [comment]
The output file does have one validation error – the `materials: []` array is empty, and empty arrays are supposed to be omitted. That would be solved if we fixed #15351, or could be solved separately, and may explain why the model doesn't load in Oculus Home. Or it could be a bug in Oculus Home, too.

**Comment by marcocorriero** [comment]
> I tried loading the .glb in the editor, then exporting it, then loading it again and I can see the model (minus the animations and material).
> 
> If the .glb doesn't load in Oculus home, you may want to report that to Oculus/Facebook instead.

before opening or dropping the file , i always try to open with 3D viewer from windows, and it says "impossible to open this model, try again." , this means there's a issue on exporting, maybe fix and add animation supports...

**Comment by donmccurdy** [comment]
Try this file:

[toaster.glb.zip](https://github.com/mrdoob/three.js/files/2933330/toaster.glb.zip)

That file is fully valid according to the official glTF validator, although it is still missing animation and materials. If Oculus Home or Windows 3D viewer cannot open it, they have a bug.

**Comment by marcocorriero** [comment]
![image](https://user-images.githubusercontent.com/16446639/53841202-7ca2c180-3f9c-11e9-8da7-f051559cafd3.png)


Perfectly loaded.

**Comment by donmccurdy** [comment]
Ok, I'll put a 3rd item on this list then:

- GLTFExporter: Support spec/gloss PBR materials (#15351)
- Would like animations on models imported in the editor included on scene export (#11642)
- [ ] GLTFExporter materials/meshes/lights/... array output should never be empty. 

Reopening until we address that 3rd item.

**Comment by marcocorriero** [comment]
Maybe add even a way to edit models with multiple animations, soo i could edit and remove other animations and export the same model with different animations..

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
